update CNN ROI modules

wwuoneway opened the pull request:
- update EvaluateROIEff_module.cc, more accurate evaluation on efficiency and purity;
- update RawWaveformDump_module.cc to have options use either RawDigit or WireProducer waveform, to save full waveform or short waveform;
- update WaveformRoiFinder_module.cc and waveformroifinder.fcl to use cnnroi network for each plane separately, add an option to set number of planes.

FNALbuild commented:
A new Pull Request was created by  @wwuoneway  for develop.

It involves the following packages:

larrecodnn


@LArSoft/level-1-managers, @LArSoft/level-2-managers can you please review it and eventually sign? Thanks.

cms-bot commands are listed <a href="https://cdcvs.fnal.gov/redmine/projects/larsoft/wiki/Pull_request_comments_that_trigger_CI_actions">here</a>

FNALbuild commented:
The code-checks are being triggered in jenkins.

FNALbuild commented:
+code-checks  
 for details see 
 https://buildmaster.fnal.gov/buildmaster/job/GithubIntegration/job/run-pr-code-checks//688/console

lgarren commented:
trigger build

FNALbuild commented:
The tests are being triggered in jenkins.

FNALbuild commented:
+LArSoft tests OK on slf7 for c7:prof 
 for details see 
 https://lar-ci-history.fnal.gov/LarCI/app/view_builds/index?offset=0&builds=lar_ci/9485&builds= 
 +build

FNALbuild commented:
+LArSoft tests OK on slf7 for e19:prof 
 for details see 
 https://lar-ci-history.fnal.gov/LarCI/app/view_builds/index?offset=0&builds=lar_ci/9484&builds= 
 +build

FNALbuild commented:
-lariat failed on slf7 for e19:prof 
 for details see 
 https://lar-ci-history.fnal.gov/LarCI/app/ns:lariat/view_builds/index?offset=0&builds=lariat_ci/4935&builds= 
 for details of the parent CI build see 
 https://lar-ci-history.fnal.gov/LarCI/app/view_builds/index?offset=0&builds=lar_ci/9484&builds=

FNALbuild commented:
-icarus tests failed on slf7 for e19:prof 
 for details see 
 https://lar-ci-history.fnal.gov/LarCI/app/ns:icarus/view_builds/index?offset=0&builds=icarus_ci/1085&builds= 
 for details of the parent CI build see 
 https://lar-ci-history.fnal.gov/LarCI/app/view_builds/index?offset=0&builds=lar_ci/9484&builds=

FNALbuild commented:
-argoneut tests warning on slf7 for e19:prof 
 for details see 
 https://lar-ci-history.fnal.gov/LarCI/app/ns:argoneut/view_builds/index?offset=0&builds=argoneut_ci/5120&builds= 
 for details of the parent CI build see 
 https://lar-ci-history.fnal.gov/LarCI/app/view_builds/index?offset=0&builds=lar_ci/9484&builds=

FNALbuild commented:
-uboone tests failed on slf7 for e19:prof 
 for details see 
 https://lar-ci-history.fnal.gov/LarCI/app/ns:uboone/view_builds/index?offset=0&builds=uboone_ci/5532&builds= 
 for details of the parent CI build see 
 https://lar-ci-history.fnal.gov/LarCI/app/view_builds/index?offset=0&builds=lar_ci/9484&builds=

knoepfel commented on larrecodnn/ImagePatternAlgs/Tensorflow/Modules/waveformroifinder.fcl:
```diff
@@ -3,8 +3,50 @@ BEGIN_PROLOG
     WireProducerLabel:  "caldata:dataprep"
-    WaveformRecog: {
+
+    WaveformRecogPlane0: {
```
There are a few issues here:

- The number of planes should be determined by the geometry service (as mentioned above)
- The `WaveformSize` is stated both here and in the configurations of the tools; is it necessary in both places?
- Why provide a `RawProducerLabel` default of the empty string?  That is the default value in the C++ file anyway and it is not necessary to provide it here.
- If you think it is necessary to have a tool for each plane, then it would be much safer (and simpler in the C++ file) to use a sequence of configurations instead of configurations with hard-coded indices:
    ```.jsonnet
    WaveformRecogs: [
        { /* config for first tool */ },
        { /* config for second tool */ },
        { /* config for third tool */ }
    ]
    ```
- The configurations for each of the tools appear to be identical.  So do you really need multiple tools for this module?  If you decide you do, you should introduce a FHiCL parameter that can be used elsewhere:
    ```.jsonnet
    tool_config: {
      NNetModelFile: "CnnModels/lightmodel112.pb"
      ...
    }
    # And then in the configuration body:
    WaveformRecogs: [
      @local::tool_config,
      @local::tool_config,
      @local::tool_config
    ]
    ```

knoepfel commented on larrecodnn/ImagePatternAlgs/Tensorflow/Modules/EvaluateROIEff_module.cc:
```diff
@@ -47,104 +56,433 @@ class nnet::EvaluateROIEff : public art::EDAnalyzer {
+    
+    if (signal_starttick.size() == 0 || (signal_endtick.size()==1 && signal_endtick.back()==-999)) continue;
+
+    /*
+    // scan signals
+    cout << "signal_starttick.size(): " << signal_starttick.size() << endl;
+    cout << "signal_endtick.size(): " << signal_endtick.size() << endl;
+    cout << "signal_energydeposits.size(): " << signal_energydeposits.size() << endl;
+    cout << "signal_energy_max.size(): " << signal_energy_max.size() << endl;
+    cout << "signal_max_tdctick.size(): " << signal_max_tdctick.size() << endl;
+    for (size_t s=0; s<signal_starttick.size(); s++) {
+      cout << "....signal: " <<  s << endl;
+      cout << "start: " << signal_starttick[s] << " ; end: " << signal_endtick[s] << endl;
+      cout << "energy deposits: " << signal_energydeposits[s] << endl;
+      cout << "energy maxE: " << signal_energy_max[s] << endl;
+      cout << "energy maxE tdctick: " << signal_max_tdctick[s] << endl;
+    }
+    cout << "first tdc: " << signal_starttick[0] << "; last tdc: " << signal_endtick.back() << endl;
+    */
```
Remove commented-out code.

knoepfel commented on larrecodnn/ImagePatternAlgs/Tensorflow/Modules/EvaluateROIEff_module.cc:
```diff
@@ -2,9 +2,15 @@
 // Plugin Type: analyzer (art v3_05_00)
 // File:        EvaluateROIEff_module.cc
-//
+// 
```
This PR includes a lot of whitespace-only changes.  My guess is you worked on this PR before updating to LArSoft v9?   If so,  please try to rebase this code on top of v9 so that you can avoid these formatting issues.

knoepfel commented on larrecodnn/ImagePatternAlgs/Tensorflow/Modules/EvaluateROIEff_module.cc:
```diff
@@ -47,104 +56,433 @@ class nnet::EvaluateROIEff : public art::EDAnalyzer {
+        e_deposit += energyDeposit.energy;
+      }
+
+      //std::cout<<e_deposit<<std::endl;
+
```
Remove commented-out code.

knoepfel commented on larrecodnn/ImagePatternAlgs/Tensorflow/Modules/EvaluateROIEff_module.cc:
```diff
@@ -47,104 +56,433 @@ class nnet::EvaluateROIEff : public art::EDAnalyzer {
+      
+      // loop over signals:
+      // a) if signal s is not in any ROI (including the case no ROI), fill h_energy with signal_energy_max[s];
+      // b) if signal s is in a ROI, check following signals that are also in this ROI, then use the maximum of signal_energy_max to fill h_energy and h_energy_roi. After this, the loop will skip to the signal that is not in this ROI.
```
These lines are getting pretty long.  Can you wrap that comment onto the next line(s)?

FNALbuild commented:
-sbnd tests warning on slf7 for e19:prof 
 for details see 
 https://lar-ci-history.fnal.gov/LarCI/app/ns:sbnd/view_builds/index?offset=0&builds=sbnd_ci/5475&builds= 
 for details of the parent CI build see 
 https://lar-ci-history.fnal.gov/LarCI/app/view_builds/index?offset=0&builds=lar_ci/9484&builds=

knoepfel commented on larrecodnn/ImagePatternAlgs/Tensorflow/Modules/EvaluateROIEff_module.cc:
```diff
@@ -47,104 +56,433 @@ class nnet::EvaluateROIEff : public art::EDAnalyzer {
-              << std::endl;
-    if (TEfficiency::CheckConsistency(*(h_energy[i]), *(h_energy_roi[i]))) {
+  for (int i = 0; i<3; ++i){
+    std::cout<<i<<" "<<h_energy[i]->GetNbinsX()<<" "<<h_energy_roi[i]->GetNbinsX()<<std::endl;
```
If printing out such information is necessary, please use `mf::LogInfo`, which allows a user to redirect or disable the logging of messages.

knoepfel commented on larrecodnn/ImagePatternAlgs/Tensorflow/Modules/EvaluateROIEff_module.cc:
```diff
@@ -47,104 +56,433 @@ class nnet::EvaluateROIEff : public art::EDAnalyzer {
+    }
+  }
+
+  // purity
+  for (int i=0; i<3; ++i) {
+    if (fCount_Roi_total[i]==0) continue;
+    cout << "\nplane : " << i << endl;
+    cout << "ROI purity (a) sig roi/total rois: " << fCount_Roi_sig[i] << "/" << fCount_Roi_total[i] << " = " << 1.0*fCount_Roi_sig[i]/fCount_Roi_total[i] << endl;
+
+    cout << "ROI purity (b) from histogram: sig/total: " << h_roi[i]->GetBinContent(2) << "/" << h_roi[i]->GetBinContent(1) << " = " << 1.0*(h_roi[i]->GetBinContent(2))/(h_roi[i]->GetBinContent(1)) << endl;
+  }
+
+  cout << "\nROI combined purity (all planes) (a) sig roi/total rois: " << fCount_Roi_sig[0]+ fCount_Roi_sig[1] + fCount_Roi_sig[2]<< "/" << fCount_Roi_total[0]+fCount_Roi_total[1] + fCount_Roi_total[2] << " = " << 1.0*(fCount_Roi_sig[0]+fCount_Roi_sig[1]+fCount_Roi_sig[2])/(fCount_Roi_total[0]+fCount_Roi_total[1]+fCount_Roi_total[2]) << endl;
+
+  cout << "ROI combined purity (all planes) (b) from histogram: " << h_roi[0]->GetBinContent(2) + h_roi[1]->GetBinContent(2) + h_roi[2]->GetBinContent(2) << "/" << h_roi[0]->GetBinContent(1) + h_roi[1]->GetBinContent(1) + h_roi[2]->GetBinContent(1) << " = "<< 1.0*(h_roi[0]->GetBinContent(2) + h_roi[1]->GetBinContent(2) + h_roi[2]->GetBinContent(2)) / (h_roi[0]->GetBinContent(1) + h_roi[1]->GetBin ...
```
As these lines of code are only for printing out information, I suggest removing them.

knoepfel commented on larrecodnn/ImagePatternAlgs/Tensorflow/Modules/EvaluateROIEff_module.cc:
```diff
@@ -47,104 +56,433 @@ class nnet::EvaluateROIEff : public art::EDAnalyzer {
+  //   (i) signal is totally in the ROI
+  //   (ii) signal is partially in the ROI
+  // Equivalently, we can just check whether the maxtick is in the ROI (simple one). We keep the above algorithm here, too.
+  bool IsInROI = false;
+  /*
+  // signal is totally in ROI
+  if (starttick>=roistart && endtick<roiend) {
+    IsInROI = true;
+  }
+  // signal is partial in ROI
+  else if ( (starttick<roistart && roistart<=endtick) ||
+            (starttick<roiend && roiend <= endtick) ) {
+    if (roistart<=maxtick && maxtick<roiend) {
+      IsInROI = true;
     }
   }
+  */
+
+  if (roistart<=maxtick && maxtick<roiend) {
+    IsInROI = true;
+  }
+
+  return IsInROI;
```
This entire implementation can be replaced with:
```.cpp
bool nnet::EvaluateROIEff::isSignalInROI (int starttick, int endtick, int maxtick, int roistart, int roiend) {
  return roistart <= maxtick && maxtick < roiend;
}
```

knoepfel commented on larrecodnn/ImagePatternAlgs/Tensorflow/Modules/RawWaveformDump_module.cc:
```diff
@@ -329,8 +386,8 @@ nnet::RawWaveformDump::analyze(art::Event const& evt)
         unsigned int tdctick = static_cast<unsigned int>(clockData.TPCTDC2Tick(double(tpctime)));
-        if (tdctick != tpctime)
-          std::cout << "tpctime: " << tpctime << ", tdctick: " << tdctick << std::endl;
+        //if (tdctick != tpctime)
+        //  std::cout << "tpctime: " << tpctime << ", tdctick: " << tdctick << std::endl;
```
Remove commented code.

knoepfel commented on larrecodnn/ImagePatternAlgs/Tensorflow/Modules/WaveformRoiFinder_module.cc:
```diff
@@ -76,8 +80,12 @@ nnet::WaveformRoiFinder::WaveformRoiFinder(fhicl::ParameterSet const& p)
   // Signal/Noise waveform recognition tool
-  fWaveformRecogTool =
-    art::make_tool<wavrec_tool::IWaveformRecog>(p.get<fhicl::ParameterSet>("WaveformRecog"));
+  fWaveformRecogToolVec.resize(fNPlanes);
+  for (int n=0; n<fNPlanes; n++) {
+    char planeWaveformRecog[30];
+    sprintf(planeWaveformRecog, "WaveformRecogPlane%d",n);
+    fWaveformRecogToolVec[n] = art::make_tool<wavrec_tool::IWaveformRecog>(p.get<fhicl::ParameterSet>(planeWaveformRecog));
+  }
```
Assuming you really need each plane to have different tools, I suggest a different approach--the configuration file should specify a sequence of configurations for each tool (see FHiCL file comment), and then the tools can be easily constructed using:
```.cpp
fWaveformRecogToolVec.reserve(fNPlanes);
auto const tool_psets = p.get<std::vector<fhicl::ParameterSet>>("WaveformRecogs");
for (auto const& pset : tool_psets) {
  fWaveformRecogToolVec.push_back(art::make_tool<wavrec_tool::IWaveformRecog>(pset));
}
```

knoepfel commented on larrecodnn/ImagePatternAlgs/Tensorflow/Modules/waveformroifinder.fcl:
```diff
@@ -3,8 +3,50 @@ BEGIN_PROLOG
+    }
+
+    WaveformRecogPlane1: {
+       NNetModelFile: "CnnModels/lightmodel112.pb"
```
How does the user get access to the "CnnModels/lightmodel112.pb" file?  Is it already included in a PR somewhere?  Or do they get it through CVMFS?

wwuoneway replied:
@knoepfel Thank you so much for your detailed suggestions. Those are really helpful. 
Changes are made except the one related to "CnnModels/lightmodel112.pb". This part is along with the experiment release and can get it through CVMFS.   Regarding the tool for each plane, yes, we want to keep the individual tool for each plane, though the configuration here is using the same one, which can be easily replaced in the experiment fcl file.
Let me know if there are other  issues. Again, many thanks for your comments!

FNALbuild commented:
-dune tests failed on slf7 for e19:prof 
 for details see 
 https://lar-ci-history.fnal.gov/LarCI/app/ns:dune/view_builds/index?offset=0&builds=dune_ci/10217&builds= 
 for details of the parent CI build see 
 https://lar-ci-history.fnal.gov/LarCI/app/view_builds/index?offset=0&builds=lar_ci/9484&builds=

FNALbuild commented:
Pull request #10 was updated. @LArSoft/level-1-managers, @LArSoft/level-2-managers can you please check and sign again.

FNALbuild commented:
The code-checks are being triggered in jenkins.

FNALbuild commented:
+code-checks  
 for details see 
 https://buildmaster.fnal.gov/buildmaster/job/GithubIntegration/job/run-pr-code-checks//690/console

wwuoneway commented:
@knoepfel Just let you know, I already addressed your comments. I wish I didn't miss some of them. Please let me know if it still has a issue. Thank you!

lgarren commented:
trigger build

FNALbuild commented:
The tests are being triggered in jenkins.

FNALbuild commented:
+LArSoft tests OK on slf7 for c7:prof 
 for details see 
 https://lar-ci-history.fnal.gov/LarCI/app/view_builds/index?offset=0&builds=lar_ci/9489&builds= 
 +build

FNALbuild commented:
+LArSoft tests OK on slf7 for e19:prof 
 for details see 
 https://lar-ci-history.fnal.gov/LarCI/app/view_builds/index?offset=0&builds=lar_ci/9488&builds= 
 +build

FNALbuild commented:
-lariat failed on slf7 for e19:prof 
 for details see 
 https://lar-ci-history.fnal.gov/LarCI/app/ns:lariat/view_builds/index?offset=0&builds=lariat_ci/4937&builds= 
 for details of the parent CI build see 
 https://lar-ci-history.fnal.gov/LarCI/app/view_builds/index?offset=0&builds=lar_ci/9488&builds=

FNALbuild commented:
-icarus tests failed on slf7 for e19:prof 
 for details see 
 https://lar-ci-history.fnal.gov/LarCI/app/ns:icarus/view_builds/index?offset=0&builds=icarus_ci/1087&builds= 
 for details of the parent CI build see 
 https://lar-ci-history.fnal.gov/LarCI/app/view_builds/index?offset=0&builds=lar_ci/9488&builds=

FNALbuild commented:
-uboone tests failed on slf7 for e19:prof 
 for details see 
 https://lar-ci-history.fnal.gov/LarCI/app/ns:uboone/view_builds/index?offset=0&builds=uboone_ci/5534&builds= 
 for details of the parent CI build see 
 https://lar-ci-history.fnal.gov/LarCI/app/view_builds/index?offset=0&builds=lar_ci/9488&builds=

FNALbuild commented:
-argoneut tests warning on slf7 for e19:prof 
 for details see 
 https://lar-ci-history.fnal.gov/LarCI/app/ns:argoneut/view_builds/index?offset=0&builds=argoneut_ci/5122&builds= 
 for details of the parent CI build see 
 https://lar-ci-history.fnal.gov/LarCI/app/view_builds/index?offset=0&builds=lar_ci/9488&builds=

FNALbuild commented:
-sbnd tests warning on slf7 for e19:prof 
 for details see 
 https://lar-ci-history.fnal.gov/LarCI/app/ns:sbnd/view_builds/index?offset=0&builds=sbnd_ci/5477&builds= 
 for details of the parent CI build see 
 https://lar-ci-history.fnal.gov/LarCI/app/view_builds/index?offset=0&builds=lar_ci/9488&builds=

FNALbuild commented:
-dune tests failed on slf7 for e19:prof 
 for details see 
 https://lar-ci-history.fnal.gov/LarCI/app/ns:dune/view_builds/index?offset=0&builds=dune_ci/10220&builds= 
 for details of the parent CI build see 
 https://lar-ci-history.fnal.gov/LarCI/app/view_builds/index?offset=0&builds=lar_ci/9488&builds=

yangtj207 commented:
This PR should not affect any experiments as the CNN ROI has not been integrated into any official reconstruction workflow yet.

lgarren commented:
The CI tests are mostly normal.  There is an unrelated problem with dunetpc and some experiments have not updated their code.

yangtj207 commented:
+1

FNALbuild commented:
This pull request is fully signed and it will be merged to develop and built in the next LArSoft release after it passes the integration tests. This pull request will now be reviewed by the release team before it's merged.)

knoepfel commented:
approve
